File: app/src/main/java/org/stepic/droid/web/OAuthService.java
package org.stepic.droid.web;

import androidx.annotation.Nullable;

import io.reactivex.Completable;
import io.reactivex.Single;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.POST;

public interface OAuthService {
    @FormUrlEncoded
    @POST("/oauth2/token/")
    Call<OAuthResponse> updateToken(@Field("grant_type") String grantType,
                                    @Field("refresh_token") String refreshToken);

    @FormUrlEncoded
    @POST("/oauth2/token/")
    Single<OAuthResponse> authWithLoginPassword(@Field("grant_type") String grant_type,
                                              @Field(value = "username", encoded = true) String username,
                                              @Field(value = "password", encoded = true) String password);


    @FormUrlEncoded
    @POST("/oauth2/token/")
    Single<OAuthResponse> getTokenByCode(@Field("grant_type") String grant_type,
                                       @Field("code") String code,
                                       @Field("redirect_uri") String redirectUri);

    @FormUrlEncoded
    @POST("/oauth2/social-token/")
    Single<OAuthResponse> getTokenByNativeCode(@Field("provider") String providerName,
                                             @Field("code") String providerCode,
                                             @Field("grant_type") String grantType,
                                             @Field("redirect_uri") String redirectUri,
                                             @Field("code_type") String accessToken,
                                             @Nullable
                                                            @Field("email") String email);


    @POST("/api/users")
    Completable createAccount(@Body UserRegistrationRequest user);
}

File: app/src/main/java/org/stepik/android/remote/auth/service/OAuthService.kt
package org.stepik.android.remote.auth.service

import io.reactivex.Completable
import io.reactivex.Single
import org.stepic.droid.web.UserRegistrationRequest
import org.stepik.android.remote.auth.model.OAuthResponse
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.POST

interface OAuthService {
    @FormUrlEncoded
    @POST("/oauth2/token/")
    fun updateToken(
        @Field("grant_type") grantType: String,
        @Field("refresh_token") refreshToken: String
    ): Call<OAuthResponse>

    @FormUrlEncoded
    @POST("/oauth2/token/")
    fun authWithLoginPassword(
        @Field("grant_type") grant_type: String,
        @Field(value = "username", encoded = true) username: String,
        @Field(value = "password", encoded = true) password: String
    ): Single<OAuthResponse>

    @FormUrlEncoded
    @POST("/oauth2/token/")
    fun getTokenByCode(
        @Field("grant_type") grant_type: String,
        @Field("code") code: String,
        @Field("redirect_uri") redirectUri: String
    ): Single<OAuthResponse>

    @FormUrlEncoded
    @POST("/oauth2/social-token/")
    fun getTokenByNativeCode(
        @Field("provider") providerName: String,
        @Field("code") providerCode: String,
        @Field("grant_type") grantType: String,
        @Field("redirect_uri") redirectUri: String,
        @Field("code_type") accessToken: String?,
        @Field("email") email: String?
    ): Single<OAuthResponse>

    @POST("/api/users")
    fun createAccount(@Body user: UserRegistrationRequest): Completable
}